src/net/http.rs

```diff
@@ -0,0 +1,1526 @@
+                preamble.consensus_serialize()
+            },
+            HttpRequestType::GetBlock(ref md, ref block_hash) => {
+                let preamble = HttpRequestPreamble::new("GET".to_string(), format!("/v2/blocks/{}", block_hash.to_hex()), md.peer.hostname(), md.peer.port());
+                preamble.consensus_serialize()
+            },
+            HttpRequestType::GetMicroblocks(ref md, ref block_hash) => {
+                let preamble = HttpRequestPreamble::new("GET".to_string(), format!("/v2/microblocks/{}", block_hash.to_hex()), md.peer.hostname(), md.peer.port());
+                preamble.consensus_serialize()
+            }
+            HttpRequestType::PostTransaction(ref md, ref tx) => {
+                let mut tx_bytes = tx.consensus_serialize();
+                let mut preamble = HttpRequestPreamble::new("POST".to_string(), "/v2/transactions".to_string(), md.peer.hostname(), md.peer.port());
+                preamble.set_content_length(tx_bytes.len() as u32);
+                preamble.set_content_type(HttpContentType::Bytes);
+
+                let mut bytes = preamble.consensus_serialize();
+                bytes.append(&mut tx_bytes);
+                bytes
+            }
+        };
+        bytes
+    }
+
+    fn consensus_deserialize(buf: &[u8], index_ptr: &mut u32, max_len: u32) -> Result<HttpRequestType, net_error> {
+        let mut index = *index_ptr;
+        let http_preamble = HttpRequestPreamble::consensus_deserialize(buf, &mut index, max_len)?;
+
+        let max_http_body_len = max_len.checked_sub(index).expect("FATAL: somehow read beyond end of buffer");
+        if http_preamble.get_content_length() > max_http_body_len {
+            return Err(net_error::OverflowError("Would read beyond end of buffer to read HTTP request body".to_string()));
+        }
+        if (http_preamble.get_content_length() as usize) > buf[(index as usize)..].len() {
+            return Err(net_error::UnderflowError("Not enough bytes to read HTTP resposne body".to_string()));
+        }
+
+        let http_body = HttpRequestType::parse(&http_preamble, &buf[(index as usize)..])?;
+        *index_ptr = index + http_preamble.get_content_length();
+        Ok(http_body)
+    }
+}
+
+impl StacksMessageCodec for HttpResponseType {
+    fn consensus_serialize(&self) -> Vec<u8> {
+        let bytes = match *self {
+            HttpResponseType::Neighbors(ref md, ref neighbor_data) => {
+                let mut neighbor_bytes = match serde_json::to_string(neighbor_data) {
+                    Ok(neighbor_str) => neighbor_str.into_bytes(),
+                    Err(_) => {
+                        return self.error_response(500, "Failed to serialize neighbor data to JSON");
+                    }
+                };
+                let mut preamble = HttpResponsePreamble::new(200, "OK".to_string(), neighbor_bytes.len() as u32, HttpContentType::JSON, md.request_id, md.request_path.clone());
```

Comment:
`consensus_serialize`, as designed, forces excessive buffering/cloning. This happens in nearly all of the functions that implement the trait, but these lines provide a fairly straight-forward case.
Line 955 serializes neighbor_data to a String -- that's one data clone.
Line 961 instantiates 1 struct, with more or less all of the data in the response.
Line 964 serializes that struct to a buffer -- perhaps this a necessary buffer, perhaps not, it depends on the caller.

That's three representations of data that could just all write directly to a byte stream.